subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/artifactreuse/ArtifactResolutionQueryIntegrationTest.groovy

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.resolve.artifactreuse
+
+import org.gradle.integtests.fixtures.AbstractHttpDependencyResolutionTest
+import org.gradle.test.fixtures.server.http.BlockingHttpServer
+import spock.lang.Issue
+import spock.lang.Timeout
+
+import java.util.concurrent.TimeUnit
+
+class ArtifactResolutionQueryIntegrationTest extends AbstractHttpDependencyResolutionTest {
+    BlockingHttpServer server = new BlockingHttpServer()
+
+    def setup() {
+        server.start()
+    }
+
+    def cleanup() {
+        server.stop()
+    }
+
+    @Issue('https://github.com/gradle/gradle/issues/3579')
+    @Timeout(value = 10, unit = TimeUnit.SECONDS)
+    def 'can use artifact resolution queries in parallel to file resolution'() {
+        given:
+        def module = mavenHttpRepo.module('group', "artifact", '1.0').publish()
+        server.expectConcurrent(server.file(module.pom.path, module.pom.file), server.resource('/sync'))
```

Comment:
❌ I don't think this is testing the race reliably - We need to make sure that both are trying to hit the cache at the same time. You can use some of the [blocking methods](https://github.com/gradle/gradle/blob/master/subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/BlockingHttpServer.java#L201) to ensure that both tasks are actually resolving at the same time.

Reply:
I guess you're misled by the name `expectConcurrent`. This is a blocking method:

```
/**
     * Expects the given requests to be made concurrently. Blocks each request until they have all been received then releases them all.
     */
```

Reply:
It waits for the `resolve` method and the `sync` call. But the `sync` call happens before the artifact resolution query. So there is no guarantee that the resolution query will be executed in parallel to `resolve`. `resolve` could finish before that. The test would be green, even if the problem isn't fixed.

I think we need to change the query so it looks for some other component first - that way we know that the query has started (and in the old code that meant the cache lock was held). Then we can block the query until the resolve method has started. That should reliably reproduce the old deadlock.

Reply:
Yes, you're right. Fixed as you say. Thanks for pointing out this.